Add a few more thinblock tests and checks

ptschip opened the pull request.

ptschip committed:
Add IsThinBlockValid() when processing thinblock or xthinblock

Check whether there are any transactions in the thinblock and
whether the first transaction is a coinbase.

Add CheckBlockHeader() to IsThinBlockValid()

Remove dead code for CheckBlockHeader in CXThinblock

Now that we have IsThinBlockValid() we no longer need nor were
using CheckBlockHeader in CXThinblock.

Add thinblock test for invalid block header

ptschip committed:
Add message consistency check for xblocktx and get_xblocktx

Make sure the messages have the basic data that they should
contain, such as a blockhash and a non-empty set of txns
or hashes to request.

ptschip committed:
Simplify and condense the consistency checking for xblocktx

Switch to Allman bracing for xblocktx

Return false if consistency check fails for xblocktx

ptschip committed:
Add two more consistency checks for get_xthin

Add a check for the bloom filter size as well as whether the block
hash is NULL.

Also, tidy up the error messages and formatting.

ptschip committed:
consistent use of error() message

use error() when returning false rather than a LogPrintf followed by
a return.  It looks better and uses one less line of code.

ptschip commented:
@gandrewstone updated to allman formatting but put into WIP as I'd like to add a few more unit tests

ptschip committed:
Add testcases for message consistency checking

Add tests for XBLOCKTX, GET_XTHIN and GET_XBLOCKTX

Took out check for IsWithinSizeConstraint() in GET_XBLOCKTX because
that is already handled in FilterLoad() and therefore is redundant.

ptschip commented:
@gandrewstone taking out of WIP.  I've got all the unit tests covered now for the xthin p2p messages.

sickpig commented on src/main.cpp:
```diff
@@ -5978,19 +5981,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             BlockMap::iterator mi = mapBlockIndex.find(inv.hash);
-            if (mi == mapBlockIndex.end()) {  // This block does not even exist
-                LogPrint("thin", "Peer %s (%d) requested nonexistent block %s\n", pfrom->addrName.c_str(), pfrom->id, inv.hash.ToString());
+            if (mi == mapBlockIndex.end())  // This block does not even exist
```
minor nit pick I would move the trialing comment up

sickpig commented on src/test/thinblock_tests.cpp:
```diff
@@ -116,6 +126,231 @@ BOOST_AUTO_TEST_CASE(thinblock_test) {
+    CNode::ClearBanned();
+    CDataStream vRecv1(SER_NETWORK, PROTOCOL_VERSION);
+    CXThinBlock xthin = xthinblock;
+    xthin.vMissingTx.clear(); // empty the missingtx vector. This should cause an error.
```
very minor formatting nit: this change set mixed trailing comments with comments on separated lines. As stated already I prefer the latter, but at least we should have a consistent approach.

sickpig reviewed (commented):
Other than the formatting nits, I've run the new test locally and it worked as intended.

ptschip committed:
Take out redundant comment

The error messeage already says what the problem is we
don't need another code comment to inform the dev of what is
happening here.

ptschip commented:
On 05/04/2017 1:26 AM, Andrea Suisani wrote:
>
> *@sickpig* commented on this pull request.
>
> Other than the formatting nits, I've run the new test locally and it
> worked as intended.
>
> ------------------------------------------------------------------------
>
> In src/main.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/406#discussion_r109853635>:
>
> > @@ -5978,19 +5981,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
>          {
>              LOCK(cs_main);
>              BlockMap::iterator mi = mapBlockIndex.find(inv.hash);
> -            if (mi == mapBlockIndex.end()) {  // This block does not even exist
> -                LogPrint("thin", "Peer %s (%d) requested nonexistent block %s\n", pfrom->addrName.c_str(), pfrom->id, inv.hash.ToString());
> +            if (mi == mapBlockIndex.end())  // This block does not even exist
>
> minor nit pick I would move the trialing comment up
>

I just took it out, it's basically a duplicate because it already tells
you in the error message what is happening, we don't need another code
comment here.
> ------------------------------------------------------------------------
>
> In src/test/thinblock_tests.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/406#discussion_r109856183>:
>
> > +       a ban will result.
> +     */
> +    fImporting = false;
> +    fReindex = false;
> +
> +    /** XTHINBLOCK message consistency checks */
> +
> +    // testing empty missingtx vector
> +    CAddress addr1(ipaddress(0xa0b0c001));
> +    CNode dummyNode1(INVALID_SOCKET, addr1, "", true);
> +    dummyNode1.nVersion = 1;
> +
> +    CNode::ClearBanned();
> +    CDataStream vRecv1(SER_NETWORK, PROTOCOL_VERSION);
> +    CXThinBlock xthin = xthinblock;
> +    xthin.vMissingTx.clear(); // empty the missingtx vector. This should cause an error.
>
> very minor formatting nit: this change set mixed trailing comments
> with comments on separated lines. As stated already I prefer the
> latter, but at least we should have a consistent approach.
>
> —
>

IDK why there's an issue with mixing, I hate to add a line and a
whitespace for some simple comment which refers to just one line of
code.  The comments on seperate lines encompass a larger set of code and
the trailing ones refer to just one line...anyway  wasn't following all
the details of what you guys came up with last week regarding the code
format.  If we're not mixing and matching anymore let me know and I'll
change it...

> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/406#pullrequestreview-30980724>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0AsWC2UQ37E3hecDDkptMgQ9JByAks5rs1BPgaJpZM4Mq-rX>.
>

zander commented on src/main.cpp:
```diff
@@ -5951,24 +5951,27 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-	      {
-		if (pfrom->nGetXthinCount >= 20)
+            if (chainparams.NetworkIDString() == "main") // other networks have variable mining rates
+	    {
```
alignment is wrong, missing space.

ptschip replied:
ok done.

zander commented on src/main.cpp:
```diff
@@ -6038,15 +6040,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-            return false;
+
+        // Check that that there is at least one txn in the xthin and that the first txn is the coinbase
+        if(!IsThinBlockValid(pfrom, thinBlock.vMissingTx, thinBlock.header))
```
missing space between if and open brace.

ptschip replied:
thank you.

zander commented on src/main.cpp:
```diff
@@ -6038,15 +6040,18 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-            Misbehaving(pfrom->GetId(), 20);
-            return false;
+
+        // Check that that there is at least one txn in the xthin and that the first txn is the coinbase
```
Why comment on what the IsThinBlockValid does? If that method changes you'd have to update this doc in 3 places.

ptschip replied:
good point, i changed the comments all to // Message Consistency checking.

zander commented on src/thinblock.cpp:
```diff
@@ -1101,6 +1101,36 @@ void SendXThinBlock(CBlock &block, CNode *pfrom, const CInv &inv)
 
+bool IsThinBlockValid(const CNode* pfrom, const std::vector<CTransaction>& vMissingTx, const CBlockHeader& header)
+{
+    // Check that that there is at least one txn in the xthin and that the first txn is the coinbase
```
Move this to be API doc (3 slashes) and move it just above the method definition.

ptschip replied:
it's fine, it returns if empty..., but i'll change it to .empty() rather than checking for size 0...that will be more clear.

zander commented on src/main.cpp:
```diff
@@ -6087,6 +6100,14 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+        {
+            LOCK(cs_main);
+            Misbehaving(pfrom->GetId(), 100);
+            return error("incorrectly constructed get_xblocktx received.  Banning peer=%d\n", pfrom->id);
```
That \n at the end is wrong, please remove.

ptschip replied:
good one..they used to be logprints, that's they got in there..thanks.  (same for all )

zander commented on src/main.cpp:
```diff
@@ -6104,9 +6125,9 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-                LogPrintf("DOS: Misbehaving - requesting too many xblocktx: %s\n", inv.hash.ToString());
                 LOCK(cs_main);
                 Misbehaving(pfrom->GetId(), 100);  // If they exceed the limit then disconnect them
+                return error("DOS: Misbehaving - requesting too many xblocktx: %s\n", inv.hash.ToString());
```
\n not wanted here.

zander commented on src/main.cpp:
```diff
@@ -6147,9 +6168,28 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+        if (thinBlockTx.vMissingTx.empty() || thinBlockTx.blockhash.IsNull() || pfrom->xThinBlockHashes.size() != pfrom->thinBlock.vtx.size())
+        {
+            {
+                LOCK(cs_vNodes);
```
Why this lock?

ptschip replied:
we need this because vnodes is not locked and since we are doing parallel validations we could have contention here .

zander replied:
What are you touching in that scope that would need locking?

ptschip replied:
mapThinBlocksInFlight

zander commented on src/main.cpp:
```diff
@@ -6147,9 +6168,28 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
+
+            LOCK(cs_main);
+            Misbehaving(pfrom->GetId(), 100);
+            return error("incorrectly constructed xblocktx or inconsistent thinblock data received.  Banning peer=%d\n", pfrom->id);
```
remove the \n in error()

zander commented on src/main.cpp:
```diff
@@ -6162,35 +6202,23 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-                pfrom->mapThinBlocksInFlight.erase(inv.hash);
-                pfrom->thinBlockWaitingForTxns = -1;
-                pfrom->thinBlock.SetNull();
+	        std::map<uint64_t, CTransaction>::iterator val = mapMissingTx.find(pfrom->xThinBlockHashes[i]);
```
May I suggest using `auto` ?

```
auto val =
```

Leaves you with more space to write longer and more useful variable names than `val` ;)

ptschip replied:
personally i'm not a big fan of auto in these situations...i like it in for loops but not my preference here...unless it was really really a long line of code.

zander commented on src/main.cpp:
```diff
@@ -6162,35 +6202,23 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
         int count=0;
         size_t i;
```
It would probably be safer if you moved the definition of `i` into the for loop, that avoids reusage later when its scope is not that useful anymore.

ptschip replied:
thank you ..done.

zander commented on src/thinblock.cpp:
```diff
@@ -1101,6 +1101,36 @@ void SendXThinBlock(CBlock &block, CNode *pfrom, const CInv &inv)
+            header.GetHash().ToString(), pfrom->addrName.c_str(), pfrom->id);
+        return false;
+    }
+    if(!vMissingTx[0].IsCoinBase())
```
A perfect 50% of the ifs have a space before the brace, the others don't... Consistency is useful to have :)

ptschip replied:
ok, fixed.

zander commented on src/thinblock.cpp:
```diff
@@ -1101,6 +1101,36 @@ void SendXThinBlock(CBlock &block, CNode *pfrom, const CInv &inv)
+bool IsThinBlockValid(const CNode* pfrom, const std::vector<CTransaction>& vMissingTx, const CBlockHeader& header)
+{
+    // Check that that there is at least one txn in the xthin and that the first txn is the coinbase
+    if(vMissingTx.size() <= 0)
```
This is wrong, you need at least one transaction, not zero.  This will cause the next if() to dereference non existing transaction.

ptschip replied:
it's fine, it returns if empty..., but i'll change it to .empty() rather than checking for size 0...that will be more clear.

zander commented on src/thinblock.cpp:
```diff
@@ -1101,6 +1101,36 @@ void SendXThinBlock(CBlock &block, CNode *pfrom, const CInv &inv)
+    }
+
+    // check block header
+    CValidationState state;
```
Using a state here makes me wonder why the other errors above were done using LogPrintf instead of using a state too.
Since 100% of the usages end up doing an `error()` call on failure, it would likely be easier to pinpoint the issue if you just used the CValidationState to throw the issue up. That also means you can avoid passing in the CNode.

ptschip replied:
actually i think we should get rid of the logprints and return error() instead...less code a more clear.

zander reviewed (commented):
Found at least one bug, and various small nags.

ptschip committed:
Some formatting and comment changes in main.cpp for thinblocks

ptschip committed:
Fix pointer alignment for IsThinBlockValid()

ptschip committed:
Fix formatting issues in thinblock.cpp for IsThinBlockValid()

Also rather than check for size 0 used .empty()

ptschip committed:
Use return error() instead of LogPrintf() in IsThinBlockValid()

This saves a line of code and is more clear.

ptschip committed:
small format fix

gandrewstone commented:
I'm getting a repeatable failure of bip68-sequence.py:
Running test sequence-lock-confirmed-inputs
Running test BIP68 not consensus before versionbits activation
Verifying nVersion=2 transactions aren't standard
Activating BIP68 (and 112/113)
[432, 336]
[432, 336]
[432, 336]

ptschip commented:
are there any more details in the logs?

On 10/04/2017 6:15 PM, gandrewstone wrote:
>
> I'm getting a repeatable failure of bip68-sequence.py:
> Running test sequence-lock-confirmed-inputs
> Running test BIP68 not consensus before versionbits activation
> Verifying nVersion=2 transactions aren't standard
> Activating BIP68 (and 112/113)
> [432, 336]
> [432, 336]
> [432, 336]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/406#issuecomment-293121895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0J2HKznvp68jDR9ssYx9yl0W-kPZks5rutQ6gaJpZM4Mq-rX>.
>
